Tf/dem error

tfuhrmann opened the pull request:
Implemented DEM error estimation generating correction values for each interferogram in subfolder /out/dem_error as well as a /out/dem_error.tif file giving the average height difference w.r.t. the input DEM.

ToDo:
- further testing of the validity of resulting correction values
- implement tests for the new functionality (i.e. dem_error.py, geometry.py and read_lookup_table in gamma.py) once the new unit test data set has been established

basaks commented on pyrate/core/dem_error.py:
```diff
@@ -14,299 +14,285 @@
+            # write dem error and correction values to file
+            mpiops.run_once(_write_dem_errors, ifg_paths, params, preread_ifgs, tiles)
+            shared.save_numpy_phase(ifg_paths, params)
+
```
It's possbile to use the `tiles_split` function here. In order to do that you may need to look into tiled or IfgPart equivalent of the `Geometry` class.

basaks commented on tests/test_dem_error.py:
```diff
@@ -0,0 +1,74 @@
+
+# compare the GAMMA and PyRate Bperp estimates
+def test_pyrate_bperp_matches_gamma_bperp(gamma_bperp):
+    np.testing.assert_array_almost_equal(pyrate_bperp / 1e6, gamma_bperp / 1e6)  # max difference < 10mm
```
@tfuhrmann, Is this something you had in mind?
From here, we just need to make sure that `pyrate_bperp` is coming from pyrate code and not as in this array `pyrate_bperp` above.

tfuhrmann commented:
> Yes I agree that we need to roll back to Thomas' version of the `remove_file_if_exists` function. Also, I wonder if `remove_file_if_exists` function is better housed in shared.py? it's generic and used in two places (dem_error.py and geometry.py)

I've now changed `remove_file_if_exists` back to the old version and moved it into shared.py

tfuhrmann commented:
I think the Python 3.6 Travis checks fail because of these lines of code:
```
    # get azimuth angle value of PyRate file azimuth_angle.tif
    out = run(f"gdallocationinfo {tif_file} {x0} {y0}", shell=True, capture_output=True,
              text=True).stdout
    angle_value = float(out.split('Value:')[1].strip('\n'))
```
Looks like `capture_output=True` is only for Python version 3.7+ (see https://stackoverflow.com/questions/52219836/subprocess-run-doesnt-return-stdout-or-stderr)

tfuhrmann commented on pyrate/core/geometry.py:
```diff
@@ -92,9 +108,17 @@ def calc_radar_coords(ifg, params, xmin, xmax, ymin, ymax):
+    :param lon: Ground point longitude for each pixel (decimal degrees).
+    :param epsilon:
+    :param heading: Satellite flight heading (radians).
+    :param roll: Right or left look direction w.r.t. satellite heading; +ve = right looking (radians).
```
I don't think 'roll' is a good choice for this parameter name. The roll angle in aviation describes a motion about the longitudinal axis (i.e. the flight direction axis), which is not what we have here. What about 'look_direction' or 'look_dir'?

See also ESA Glossary here https://earth.esa.int/handbooks/asar/CNTR5-2.html on the SAR look direction:
The radar look direction defines the angle in the horizontal plane in which the radar antenna is pointing when transmitting a pulse and receiving the return signal from the ground or from an object. The look direction is an angular measurement (in degrees) and is usually made with respect to true North. In side-looking imaging radar (SLAR), the look direction is often orthogonal (normal) to the flight trajectory (azimuth) of the platform carrying the radar and is thus synonymous with the range direction. The radar look direction is an important parameter when analysing features with a preferred orientation, for example fracture patterns in rock formations, regular street patterns, or ocean waves, as these may be enhanced through choice of appropriate radar illumination direction. ( See also "Imaging Geometry" in the geometry glossary ).

mcgarth replied:
I think your suggestion of `look_dir` is much better than my `roll`!

tfuhrmann commented on pyrate/core/geometry.py:
```diff
@@ -128,19 +162,21 @@ def calc_pixel_geometry(ifg, rg, lon, lat, params):
     lat = np.radians(lat)
     heading = np.radians(heading)
-    azimuth = np.radians(azimuth)
+    right_or_left_look = np.radians(right_or_left_look)
```
maybe replace with 'look_direction'

mcgarth replied:
I prefer `look_dir`, as in above comment

mcgarth reviewed (approved):
I have done some thorough testing of this branch over a period of about a ~week, including as it evolves over that period. I have committed minor small changes to improve the "user interface" with the new functionality. But my opinion is that the core new functionality to estimate DEM errors (including geometry and perpendicular baseline calculations) is well put together and really well documented. It is a great addition to the PyRate software - well done @tfuhrmann !
